File: prototype/rails/src/app/domain/listings/listing_draft.rb
require_relative "../money"

module Domain
  module Listings
    # The listing fields a seller types into the create or edit form, checked
    # and converted. Status, slug, and image are the portal's to decide, so a
    # draft never carries them.
    class ListingDraft < Data.define(:title, :description, :medium, :dimensions, :price, :quantity)
      LINE_LIMIT = 255
      DESCRIPTION_LIMIT = 5_000
      QUANTITY_LIMIT = 999
      DOLLARS = /\A\d+(\.\d{1,2})?\z/
      WHOLE_NUMBER = /\A\d+\z/
      IMAGE_CONTENT_TYPE = %r{\Aimage/}

      # Submitted fields, as the form sends them: strings, plus the content
      # type of the upload when there is one.
      def self.errors_for(fields)
        {
          title: title_error(fields[:title]),
          description: line_error(fields[:description], DESCRIPTION_LIMIT, "description"),
          medium: line_error(fields[:medium], LINE_LIMIT, "medium"),
          dimensions: line_error(fields[:dimensions], LINE_LIMIT, "dimensions"),
          price: price_error(fields[:price]),
          quantity: quantity_error(fields[:quantity]),
          image: image_error(fields[:image_content_type])
        }.compact
      end

      def self.from(fields)
        new(
          title: fields[:title].to_s.strip,
          description: written(fields[:description]),
          medium: written(fields[:medium]),
          dimensions: written(fields[:dimensions]),
          price: Money.from_dollars(fields[:price]),
          quantity: fields[:quantity].to_i
        )
      end

      def self.title_error(value)
        return "Enter a title." if value.to_s.strip.empty?

        line_error(value, LINE_LIMIT, "title")
      end

      def self.line_error(value, limit, field)
        "Keep the #{field} under #{limit} characters." if value.to_s.strip.length > limit
      end

      def self.price_error(value)
        "The price is an amount in dollars, like 249.00." unless DOLLARS.match?(value.to_s.strip)
      end

      def self.quantity_error(value)
        quantity = value.to_s.strip
        return if WHOLE_NUMBER.match?(quantity) && quantity.to_i <= QUANTITY_LIMIT

        "The quantity is a whole number from 0 to #{QUANTITY_LIMIT}."
      end

      def self.image_error(content_type)
        return if content_type.nil? || IMAGE_CONTENT_TYPE.match?(content_type)

        "Upload an image file."
      end

      def self.written(value)
        text = value.to_s.strip

        text.empty? ? nil : text
      end

      private_class_method :title_error, :line_error, :price_error, :quantity_error, :image_error, :written

      def attributes
        {
          title: title,
          description: description,
          medium: medium,
          dimensions: dimensions,
          price_cents: price.cents,
          quantity: quantity
        }
      end
    end
  end
end

File: prototype/node/src/app/core/listings/listing-draft.ts
import { isDollarAmount, parseDollars, type Cents } from '../money.ts'
import type { ImageFormat } from './image-format.ts'

const LINE_LIMIT = 255
const DESCRIPTION_LIMIT = 5_000
const QUANTITY_LIMIT = 999
const WHOLE_NUMBER_PATTERN = /^\d+$/

/** What an uploaded image sniffed as: a recognized format, `'unrecognized'`
 * for bytes that matched none of them, or `null`/`undefined` for no upload
 * at all. Never the browser's filename or `Content-Type` header. */
export type UploadedImageFormat = ImageFormat | 'unrecognized'

export type ListingDraftFields = {
  title?: string
  description?: string
  medium?: string
  dimensions?: string
  price?: string
  quantity?: string
  imageFormat?: UploadedImageFormat | null
}

export type ListingDraftErrors = Partial<
  Record<'title' | 'description' | 'medium' | 'dimensions' | 'price' | 'quantity' | 'image', string>
>

/** A listing as the portal saves it. Carries no status, slug, or image: those
 * are the portal's to decide. */
export type ListingDraft = {
  title: string
  description: string | null
  medium: string | null
  dimensions: string | null
  priceCents: Cents
  quantity: number
}

export type ListingDraftResult =
  | { ok: true; value: ListingDraft }
  | { ok: false; errors: ListingDraftErrors }

function lineError(value: string | undefined, limit: number, field: string): string | undefined {
  return (value ?? '').trim().length > limit ? `Keep the ${field} under ${limit} characters.` : undefined
}

function titleError(value: string | undefined): string | undefined {
  if ((value ?? '').trim().length === 0) {
    return 'Enter a title.'
  }
  return lineError(value, LINE_LIMIT, 'title')
}

function priceError(value: string | undefined): string | undefined {
  return isDollarAmount(value ?? '') ? undefined : 'The price is an amount in dollars, like 249.00.'
}

function quantityError(value: string | undefined): string | undefined {
  const quantity = (value ?? '').trim()
  if (WHOLE_NUMBER_PATTERN.test(quantity) && Number(quantity) <= QUANTITY_LIMIT) {
    return undefined
  }
  return `The quantity is a whole number from 0 to ${QUANTITY_LIMIT}.`
}

function imageError(imageFormat: UploadedImageFormat | null | undefined): string | undefined {
  if (imageFormat === null || imageFormat === undefined || imageFormat !== 'unrecognized') {
    return undefined
  }
  return 'Upload an image file.'
}

function draftErrors(fields: ListingDraftFields): ListingDraftErrors {
  const checked: readonly [keyof ListingDraftErrors, string | undefined][] = [
    ['title', titleError(fields.title)],
    ['description', lineError(fields.description, DESCRIPTION_LIMIT, 'description')],
    ['medium', lineError(fields.medium, LINE_LIMIT, 'medium')],
    ['dimensions', lineError(fields.dimensions, LINE_LIMIT, 'dimensions')],
    ['price', priceError(fields.price)],
    ['quantity', quantityError(fields.quantity)],
    ['image', imageError(fields.imageFormat)],
  ]

  const errors: ListingDraftErrors = {}
  for (const [field, message] of checked) {
    if (message !== undefined) errors[field] = message
  }

  return errors
}

function written(value: string | undefined): string | null {
  const text = (value ?? '').trim()
  return text.length === 0 ? null : text
}

/**
 * The submitted form as a draft, or every field that is wrong. The price is
 * checked and converted against one grammar (`isDollarAmount` and
 * `parseDollars` share it), so the amount the `ok` arm carries is one
 * `parseDollars` could not have refused.
 */
export function parseListingDraft(fields: ListingDraftFields): ListingDraftResult {
  const errors = draftErrors(fields)
  if (Object.keys(errors).length > 0) {
    return { ok: false, errors }
  }

  return {
    ok: true,
    value: {
      title: (fields.title ?? '').trim(),
      description: written(fields.description),
      medium: written(fields.medium),
      dimensions: written(fields.dimensions),
      priceCents: parseDollars(fields.price ?? ''),
      quantity: Number(fields.quantity),
    },
  }
}
